tech-specs/src/explore/design/ledger/pos-integration.md

```diff
@@ -0,0 +1,257 @@
+    (None, Some(post)) => {
+      // - check that any other sub-keys for this validator address didn't exist
+      // in a pre-state
+      // - check that a consensus key record is also created
+      // - check that the `post` state is set correctly
+    },
+    (Some(pre), Some(post)) => {
+      // - check that a validator has been correctly deactivated or reactivated
+    },
+    _ => false,
+  }
+  ```
+- `validator/{validator_address}/total_deltas`:
+  - find the difference between the pre-state and post-state values and add it to the `validator_total_deltas` accumulator
+- `validator/{validator_address}/voting_power`:
+  - find the post-state value and insert it into the `validator_voting_powers` accumulator
+- `slash/{validator_address}`:
+  - find the newly evidence(s), ensure that they're not already present in the pre-state, validate them and add them to the `slashes` accumulator
+  - check that the validator's total deltas have been slashed correctly
+  - check that the validator's voting power has been adjusted correctly
+  - check that the total voting power has been adjusted correctly
+  - check that the the active and inactive validator sets are correct
+  - check that if there's was a `parameters` change in any epoch after `n` in pre-state, more than 2/3 of validators' voting power still agree to the change, otherwise the queued up change must have been removed in the post-state
+- `bond/{bond_source}/{bond_validator}/delta`:
+  - for each difference between the post-state and pre-state values:
+    - if the difference is not in epoch `n` or `n + pipeline_length`, panic
+    - add it to the `bond_deltas` accumulator
+- `unbond/{unbond_source}/{unbond_validator}/deltas`:
+  - for each difference between the post-state and pre-state values:
+    - if the difference is not in epoch `n` or `n + unboding_length`, panic
+    - add it to the `unbond_deltas` accumulator
+- `validator_set/active`:
+  - set the accumulator `validator_set_changed` to true in each epoch in which it has changed
+- `validator_set/inactive`:
+  - set the accumulator `validator_set_changed` to true in each epoch in which it has changed
+
+No other storage key changes are permitted by the VP.
+
+After the storage keys iteration, we check the accumulators:
+
+- the `validator_total_deltas` must be equal to the amounts in `bond_deltas` added with the amounts in `unbond_deltas` with slashes applied where the slash's epoch is in bond's or unbond's epoch range
+- the sum of differences between the `validator_total_deltas` in epoch `n + unbonding_length` and epoch `n + pipeline_length` must be equal to the difference in PoS account's token balance between post-state and pre-state
+- for each `validator_voting_powers`:
+  - find the validator's total bonded tokens from post-state
+  - the total bonded tokens divided by `votes_per_token` must be equal to the voting power
+- for each slash in `slashes`, the `validator_total_deltas` must be reduced by the slash rate
```

Comment:
we can just check that it is reduced by the multiplicand (product) of the slash fractions

Reply:
updated in ac6bad9